mirgecom/prometheus.py

```diff
@@ -0,0 +1,578 @@
+        s_low = (
+            4.198641e00 * tt6
+            - 2.036434e-03 * tt0
+            + 6.520402e-06 * 0.50 * tt1
+            - 5.487971e-09 * self.one_third * tt2
+            + 1.771978e-12 * 0.25 * tt3
+            - 8.490322e-01
+        )
+        sr4 = actx.np.where(tt0 < 1.000000e03, s_low, s_high)
+
+        s_high = (
+            3.337279e00 * tt6
+            - 4.940247e-05 * tt0
+            + 4.994568e-07 * 0.50 * tt1
+            - 1.795664e-10 * self.one_third * tt2
+            + 2.002554e-14 * 0.25 * tt3
+            - 3.205023e00
+        )
+        s_low = (
+            2.344331e00 * tt6
+            + 7.980521e-03 * tt0
+            - 1.947815e-05 * 0.50 * tt1
+            + 2.015721e-08 * self.one_third * tt2
+            - 7.376118e-12 * 0.25 * tt3
+            + 6.830102e-01
+        )
+        sr5 = actx.np.where(tt0 < 1.000000e03, s_low, s_high)
+
+        s_high = (
+            2.926640e00 * tt6
+            + 1.487977e-03 * tt0
+            - 5.684760e-07 * 0.50 * tt1
+            + 1.009704e-10 * self.one_third * tt2
+            - 6.753351e-15 * 0.25 * tt3
+            + 5.980528e00
+        )
+        s_low = (
+            3.298677e00 * tt6
+            + 1.408240e-03 * tt0
+            - 3.963222e-06 * 0.50 * tt1
+            + 5.641515e-09 * self.one_third * tt2
+            - 2.444854e-12 * 0.25 * tt3
+            + 3.950372e00
+        )
+        sr6 = actx.np.where(tt0 < 1.000000e03, s_low, s_high)
+
+        return make_obj_array([sr0, sr1, sr2, sr3, sr4, sr5, sr6])
+
+    def get_species_gibbs(self, temperature):
+
+        h0rt = self.get_species_enthalpies(temperature)
+        s0r = self.get_species_entropies(temperature)
+        return h0rt - s0r
+
+    def get_equilibrium_constants(self, temperature):
+
+        actx = temperature.array_context
+        rt = self.gas_constant * temperature
+        c0 = actx.np.log(self.one_atm / rt)
+
+        g0rt = self.get_species_gibbs(temperature)
+
+        k_eq1 = c0 + (g0rt[2]) - (g0rt[3] + g0rt[1])
+        k_eq2 = c0 + (g0rt[4]) - (g0rt[5] + g0rt[1])
+
+        return make_obj_array([k_eq1, k_eq2])
+
+    def get_temperature(self, enthalpy_or_energy, t_guess, y, do_energy=False):
+
+        actx = enthalpy_or_energy.array_context
+
+        if do_energy is False:
+            pv_fun = self.get_mixture_specific_heat_cp_mass
+            he_fun = self.get_mixture_enthalpy_mass
+        else:
+            pv_fun = self.get_mixture_specific_heat_cv_mass
+            he_fun = self.get_mixture_internal_energy_mass
+
+        num_iter = 500
+        tol = 1.0e-6
+        t_i = t_guess * enthalpy_or_energy * 1.0 / enthalpy_or_energy
+
+        for iter in range(num_iter):
+            f = enthalpy_or_energy - he_fun(t_i, y)
+            j = -pv_fun(t_i, y)
+            dt = -f / j
+            t_i += dt
+            tresid = actx.np.abs(dt)
+            maxerr = self.discr.norm(tresid, np.inf)
+            if maxerr < tol:
+                break
```

Comment:
This is the one-and-only place where I needed a norm, and had trouble figuring out how to compute one with DOFArrays sans discretization.

Reply:
OIC, makes sense (as discussed). Thanks!